### ISSUE TITLE
Expose error messages in exceptions when API calls fail

### ISSUE THREAD

**Issue by casperOne** [dapplo/Dapplo.Jira#14]
Discovered originally here in #11:

https://github.com/dapplo/Dapplo.Jira/issues/11#issuecomment-277483776

When a call results in a 400/error status code, the JSON returned contains values in the `errorMessages` and/or `errors` properties:

```
{
    "errorMessages":[],
    "errors":{
        "customfield_10017":"Epic Name is required."
    }
}
```

The resulting exception should expose those error messages.

I suggest that instead of throwing a `System.Exception`, create and throw a `JiraException` (or `JiraApiException`, or something similar) which extends [`HttpRequestException`](https://msdn.microsoft.com/en-us/library/system.net.http.httprequestexception(v=vs.118).aspx)).

This exception would have a status code (inherited from `HttpRequestException`) as well as an property of type array/`List<string>`/`IReadOnlyCollection<string>` that exposes the `errorMessages` property as well as a `Dictionary<string, string>/IDictionary<string, string>/IReadOnlyDictionary<string, string>` that exposes the `errors` property on the returned JSON.

I'm recommending extending `HttpRequestException` because [that's the exception type that's thrown when a call to `EnsureSuccessStatusCode` throws an exception](https://github.com/dotnet/corefx/blob/3b251eb0632d1a1b70bdeb928c86d45b2bb2b897/src/System.Net.Http/src/System/Net/Http/HttpResponseMessage.cs#L161).

**Comment by Lakritzator** [comment]
I found your suggestion quite useful, thanks for the detailed idea.
I might have overlooked something, but as commented in #11 I was pretty much out of time.

**Comment by Lakritzator** [comment]
@casperOne I did fix some issues, and it should be better now. Did you have a look at it?

**Comment by casperOne** [comment]
I'm hoping to test it tonight or tomorrow night.

**Comment by casperOne** [comment]
This isn't working as expected.

I'm on v0.5.27 and an `HttpRequestException` is thrown.

When the following is returned as the body on an error:

    {
        "errorMessages": [
            "Only Connect add-on users with admin scope permission are allowed to override screen security."
        ],
        "errors": {
        }
    }

There's no way to get at that information.

**Comment by Lakritzator** [comment]
It probably depends on what you call, can you give me an example?

**Comment by casperOne** [comment]
The structure of the error is consistent no matter the call (this is JSON schema):

```
{
    "id": "https://docs.atlassian.com/jira/REST/schema/error-collection#",
    "title": "Error Collection",
    "type": "object",
    "properties": {
        "errorMessages": {
            "type": "array",
            "items": {
                "type": "string"
            }
        },
        "errors": {
            "type": "object",
            "patternProperties": {
                ".+": {
                    "type": "string"
                }
            },
            "additionalProperties": false
        },
        "status": {
            "type": "integer"
        }
    },
    "additionalProperties": false
}
```

As per "Error Responses" on:

https://docs.atlassian.com/jira/REST/cloud/

**Comment by Lakritzator** [comment]
Sorry for the late response, I had some personal things which took up quite a lot of my free time.
I still need to check, as in the error responses text you are pointing to is saying:

> **Most** resources will return a response body in addition to the status code. **Usually**, the JSON schema of the entity returned is the following...

I have seen cases where the error is not mappable. But most likely I have forgotten to test some cases.
So I will go through this today, have some other tickets too.

**Comment by Lakritzator** [comment]
Ah, you are updating an issue (edit issue), this is such a case, documentation here doesn't specify error information is returned:
https://docs.atlassian.com/jira/REST/cloud/#api/2/issue-editIssue

I will need to look into this, clearly there is error information returned :smile:

**Comment by Lakritzator** [comment]
Recent changes are published with 0.6.10, but there might still be some changes needed for other functions. I had to change some of the dependencies, due to bug #17 which could only be solved by doing so. I hope this doesn't break anything for you...
